### ISSUE TITLE
Regression in Postgres aggregate function data type inference/casting in Rails >= 7.0.5

### ISSUE THREAD

**Issue by midnightmonster** [rails/rails#48946]
### Steps to reproduce
```sql
-- Create Postgres database, required to reproduce
CREATE DATABASE type_inference_test;
```

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # gem "rails", github: "rails/rails", tag: "v7.0.4.3" # OK
  # gem "rails", github: "rails/rails", tag: "v7.0.5"   # Fail
  # gem "rails", github: "rails/rails", tag: "v7.0.7"   # Fail
  gem "rails", github: "rails/rails", branch: "main"    # Fail
  gem "pg"
end

require "active_record"
require "minitest/autorun"
require "logger"

# The issue appears appears with Postgres, not with SQLite. I have not tested other adapters.
# ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
# N.b., in sqlite3 .pluck("sum(decimal_column)").first returns a Float instead of a BigDecimal. Though that arguably is a bug, AFAICT it is _not_ a regression.
ActiveRecord::Base.establish_connection(adapter: "postgresql", database: "type_inference_test", encoding: "unicode")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :line_items, force: true do |t|
    t.integer :unit_count
    t.decimal :subtotal
  end
end

class LineItem < ActiveRecord::Base
end

class TypeInferenceBug < Minitest::Test
  def setup
    LineItem.create! unit_count: 5, subtotal: 47.98
    LineItem.create! unit_count: 1, subtotal: 13.13
  end

  # This works correctly
  def test_single_column
    assert_kind_of Integer, LineItem.sum(:unit_count)
    assert_kind_of Integer, LineItem.pluck("sum(unit_count)").first
    assert_kind_of BigDecimal, LineItem.sum(:subtotal)
    assert_kind_of BigDecimal, LineItem.pluck("sum(subtotal)").first
  end

  # All these fail starting in 7.0.5. ActiveRecord seems to apply the type of the last item to all numeric(?) columns.
  def test_two_column_decimal_last
    # Fails in 7.0.5-7.0.7, ok in 7.0.4.3
    assert_kind_of Integer, LineItem.pluck("sum(unit_count)", "sum(subtotal)").first.first
  end

  def test_two_column_integer_last
    # Fails in 7.0.5-7.0.7, ok in 7.0.4.3
    assert_kind_of BigDecimal, LineItem.pluck("sum(subtotal)", "sum(unit_count)").first.first
  end

  def test_same_problem_with_another_aggregate_function
    # Fails in 7.0.5-7.0.7, ok in 7.0.4.3
    assert_kind_of BigDecimal, LineItem.pluck("max(subtotal)", "max(unit_count)").first.first
  end
end

```

### Expected behavior
Type of values returned from `pluck("aggregate_function(column)",...)` depends on database's data type and does not change if additional values are also plucked. Up through Rails 7.0.4.3, this worked as expected.

### Actual behavior
When plucking sums (or other aggregate functions) of more than one numeric column, `pluck` coerces all numeric(?) values to the type of the last column, at least when using the postgresql adapter. In our use, this results in integers inconveniently cast to BigDecimals and (more perniciously) decimal values representing dollars and cents being truncated to integers—and the data type of existing columns of a `pluck` call changing based on order or from adding an additional column.

### System configuration
**Rails version**: 7.0.7

**Ruby version**: 3.1.1p18

**Comment by yahonda** [comment]
This behavior has been introduced via https://github.com/rails/rails/commit/81295f94c29645b2467b927ac0742ce485d19285

**Comment by midnightmonster** [comment]
We ran into this at `"#{day_job}"` via our extensive use of [activerecord-summarize](https://github.com/midnightmonster/activerecord-summarize), which uses `pluck` under the hood. I've made a new release of `activerecord-summarize` to work around the issue by transparently adding aliases to all aggregate function result columns, but it still seems weird for ActiveRecord to regress on this.

**Comment by kamipo** [comment]
Fixed by #52015.
